sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/VisualStudioCacheAccessor.java

```diff
@@ -0,0 +1,185 @@
+        return details;
+    }
+
+
+
+    /**
+     * Get the credential for the specified service and account name.
+     *
+     * @param serviceName the name of the service to lookup.
+     * @param accountName the account of the service to lookup.
+     * @return the credential.
+     */
+    public String getCredentials(String serviceName, String accountName) {
+        String credential;
+
+        if (Platform.isWindows()) {
+
+            try {
+                WindowsCredentialAccessor winCredAccessor =
+                        new WindowsCredentialAccessor(serviceName, accountName);
+                credential = winCredAccessor.read();
+            } catch (RuntimeException e) {
+                throw logger.logExceptionAsError(new CredentialUnavailableException(
+                        "Failed to read Vs Code credentials from Windows Credential API.", e));
+            }
+
+        } else if (Platform.isMac()) {
+
+            try {
+                KeyChainAccessor keyChainAccessor = new KeyChainAccessor(null,
+                        serviceName, accountName);
+
+                byte[] readCreds = keyChainAccessor.read();
+                credential = new String(readCreds, StandardCharsets.UTF_8);
+            } catch (RuntimeException e) {
+                throw logger.logExceptionAsError(new CredentialUnavailableException(
+                        "Failed to read Vs Code credentials from Mac Native Key Chain.", e));
+            }
+
+        } else if (Platform.isLinux()) {
+
+            try {
+                LinuxKeyRingAccessor keyRingAccessor = new LinuxKeyRingAccessor(
+                        "org.freedesktop.Secret.Generic", "service",
+                        serviceName, "account", accountName);
+
+                byte[] readCreds = keyRingAccessor.read();
+                credential = new String(readCreds, StandardCharsets.UTF_8);
+            } catch (RuntimeException e) {
+                throw logger.logExceptionAsError(new CredentialUnavailableException(
+                        "Failed to read Vs Code credentials from Linux Key Ring.", e));
+            }
+
+        } else {
+            throw logger.logExceptionAsError(
+                new CredentialUnavailableException(PLATFORM_NOT_SUPPORTED_ERROR));
+        }
+
+        if (CoreUtils.isNullOrEmpty(credential) || !isRefreshTokenString(credential)) {
+            throw logger.logExceptionAsError(
+                    new CredentialUnavailableException("Please authenticate via Azure Tools plugin in VS Code IDE."));
+        }
+        return credential;
+    }
+
+    private boolean isRefreshTokenString(String str) {
+        for (int i = 0; i < str.length(); i++) {
+            char ch = str.charAt(i);
+            if ((ch < '0' || ch > '9') && (ch < 'A' || ch > 'Z') && (ch < 'a' || ch > 'z')
+                        && ch != '_' && ch != '-' && ch != '.') {
+                return false;
+            }
+        }
+        return true;
```

Comment:
This code can be replaced by `return str.matches("^[-_.a-zA-Z0-9]+$")`

Reply:
String.matches() way of matching regex to String has a good perf cost.
String matches forces the input regex to always first get compiled via a Pattern.compile() and doing it on every call comes with a perf cost.
There is an alternative to String.matches and retain the perf, updated with that logic.